---
# Had issues with the bash script failing from Semaphore and this provides a cleaner mechanism for integrating with Uptime Kuma and give a nice web dashboard vs something like Cron
- name: SYNC Backups to NAS
  hosts: "{{ hosts }}"
  gather_facts: false
  vars:
    MOUNT_PATH: /mnt/nas_replicate_out

  tasks:

    # - name: Get the executing user's name
    #   ansible.builtin.command: whoami
    #   register: user_name_result

    # - name: Display executing user's name
    #   ansible.builtin.debug:
    #     msg: "Executing user's name: {{ user_name_result.stdout }}"

    # Succeeds when run from Semaphore but does nothing for some reason.
    # The wait wait_for_connection also just wizzes past with success even though the server is still asleep
    # When run locally it actually wakes the machine
    # No idea why it's not working from Seamphore. Attempting raw wakeonlan command instead
    - name: Wake NAS
      community.general.wakeonlan:
        mac: '2C:44:FD:1C:61:62'
        broadcast: 10.88.10.255
      # delegate_to: localhost

    # - name: Update apt and install etherwake for wakeonlan command
    #   apt:
    #     name: etherwake
    #     state: latest
    #     update_cache: true

    - name: Wait for NAS to become reachable
      ansible.builtin.wait_for_connection:

    # Seems to need a bit more time after wait_for_connection succeeds before it will be able to mount the share.
    # Probably have to wait until the services are up on the other end as opposed to just the "transport connection" being active
    - name: Buffer
      ansible.builtin.wait_for:
        timeout: 5
      delegate_to: localhost

    # Mount module requires a bunch of duplcate info that's already provided in fstab and configured on the host
    # Don't want to duplicate that complexity here so running like this instead
    # https://andreas.scherbaum.la/post/2019-12-09_ansible-just-mount-an-existing-filesystem/
    - name: Mount NAS
      command: "mount -v {{ MOUNT_PATH }}"
      become: true
      register: mount_result
      failed_when: mount_result.rc != 0 and mount_result.rc != 32

    - name: Execute rsync command
      # Semaphore runs as the maintenance user when executing plabooks, however, this fails with permissions errors when running as the same user locally. Not sure why
      #remote_user: maintenance
      #become: true # This seems to be required when running locally, not sure maintenance user does not work
      ansible.builtin.raw: |
        rsync -rtEvv --delete --exclude=".*/" /media/ext_storage/backups/ /mnt/nas_replicate_out/vector/backups_sync/
      register: rsync_result
      ignore_errors: yes

    - name: Notify uptime monitoring - success
      ansible.builtin.uri:
        url: "https://uptime-kuma.ournetwork.ca/api/push/XcfEl3F0QF?status=up&msg=OK&ping="
        method: GET
      when: rsync_result.rc == 0

    - name: Notify uptime monitoring - failure
      ansible.builtin.uri:
        url: "https://uptime-kuma.ournetwork.ca/api/push/XcfEl3F0QF?status=down&msg=OK&ping="
        method: GET
      when: rsync_result.rc != 0

    - name: Fail if rsync fails
      ansible.builtin.fail:
        msg: "Rsync failed with return code {{ rsync_result.rc }}."
      when: rsync_result.rc != 0

    # Precautionary sleep to give previous commands time to release their open files.
    # During testing of unmount it would fail if it was done too soon after previous interactions with the share
    - name: Sleep
      ansible.builtin.wait_for:
        timeout: 5
      delegate_to: localhost

    - name: Unmount NAS
      become: true
      ansible.posix.mount:
        path: "{{ MOUNT_PATH }}"
        state: unmounted      